### PR TITLE
devices: Add pvpanic support

### PR DESCRIPTION
**Pull request by up2wing**
When crash occurs in guest, we can use pvpanic device to notify through the event-monitor. This is useful to monitor the guest's running state and respond to the event.

The pvpanic device can be implemented as an ISA device or PCI device, we implemented the pvpanic-pci device for now.

### CODE REVIEW

File: vmm/src/device_manager.rs [hunk 1]
```diff
@@ -3604,6 +3616,63 @@ impl DeviceManager {
...
+        )?;
+
+        let mut node = device_node!(id, pvpanic_device);
+
+        node.resources = new_resources;
+        node.pci_bdf = Some(pci_device_bdf);
+        node.pci_device_handle = None;
+
+        // Add PvPanic Device to mmio
+        self.address_manager
```
Comment (rbradford) [inline vmm/src/device_manager.rs:3652]: Device BARs should already be added to the MMIO bus so this isn't necessary - if this device's BARs are not appearing then something else has gone wrong. This is done via `DeviceManager::add_pci_device` -> `PciBus::register_mapping`

Comment (up2wing) [reply vmm/src/device_manager.rs:3652]: Thanks for your quick reply, Rob.

**Comment by up2wing** [comment]
Hi Rob,
Could you please review this patch again? I update it as you suggested :)

@rbradford

File: vmm/src/config.rs [hunk 2]
```diff
@@ -1554,6 +1566,22 @@ impl VsockConfig {
     }
 }
 
+impl PvPanicConfig {
+    pub fn parse(device_type: &str) -> Result<Self> {
+        let mut parser = OptionParser::new();
+        parser.add_valueless("pci");
+        parser.parse(device_type).map_err(Error::ParsePvPanic)?;
+
+        if parser.is_set("pci") {
```
Comment (rbradford) [inline vmm/src/config.rs:1575]: I don't think we will ever support anything other than a PCI device so you don't need this.

**Comment by rbradford** [comment]
I think you need to update the integration tests.

**Comment by up2wing** [comment]
> I think you need to update the integration tests.

Yes, I am working on this.
Thanks a lot :)

**Comment by up2wing** [comment]
> I think you need to update the integration tests.

patches have been updated, Rob ;)
@rbradford

File: tests/integration.rs [hunk 3]
```diff
@@ -6135,6 +6145,56 @@ mod common_parallel {
...
+        let r = std::panic::catch_unwind(|| {
+            guest.wait_vm_boot(None).unwrap();
+
+            // Trigger guest a panic
+            make_guest_panic(&guest);
+
+            // Wait a while for guest
+            thread::sleep(std::time::Duration::new(5, 0));
+
+            // Ignore event check tempraly, as we need pvpanic-pci driver
```
Comment (rbradford) [inline tests/integration.rs:6180]: Why not add the panic driver to the kernel config file?

Comment (up2wing) [reply tests/integration.rs:6180]: Do you mean resources/linux-config-x86_64?
Will the vmlinux used in CI workloads build from this config?  Forgive me for not knowing this 😅

Comment (rbradford) [reply tests/integration.rs:6180]: Yes, the CI builds from that config file - there is an aarch64 equivalent too.

Comment (up2wing) [reply tests/integration.rs:6180]: Okay. I will try to fix this.
Thx 😸

File: vmm/src/config.rs [hunk 4]
```diff
@@ -98,6 +98,10 @@ pub enum Error {
     ParseTpm(OptionParserError),
     /// Missing path for TPM device
     ParseTpmPathMissing,
+    /// Failed parsing pvpanic
+    ParsePvPanic(OptionParserError),
+    /// No device type given for pvpanic
+    ParsePvPanicDeviceTypeGiven,
```
Comment (rbradford) [inline vmm/src/config.rs:104]: Leftover error

File: vmm/src/config.rs [hunk 5]
```diff
@@ -336,6 +340,13 @@ impl fmt::Display for Error {
             ParseVdpaPathMissing => write!(f, "Error parsing --vdpa: path missing"),
             ParseTpm(o) => write!(f, "Error parsing --tpm: {o}"),
             ParseTpmPathMissing => write!(f, "Error parsing --tpm: path missing"),
+            ParsePvPanic(o) => write!(f, "Error parsing --pvpanic: {}", o),
+            ParsePvPanicDeviceTypeGiven => {
```
Comment (rbradford) [inline vmm/src/config.rs:344]: This can also be removed.

File: resources/linux-config-aarch64 [hunk 6]
```diff
@@ -1402,7 +1402,9 @@ CONFIG_NVME_MULTIPATH=y
 # CONFIG_MISC_RTSX_PCI is not set
 # CONFIG_HABANA_AI is not set
 # CONFIG_UACCE is not set
-# CONFIG_PVPANIC is not set
+CONFIG_PVPANIC=y
+CONFIG_PVPANIC_MMIO=y
```
Comment (rbradford) [inline resources/linux-config-aarch64:1406]: Do you need this one? Does PCI depend on it?

**Review by rbradford — COMMENTED** [review]
Looking good - really close now!

File: resources/linux-config-aarch64 (hunk 6 above)
Comment (up2wing) [reply resources/linux-config-aarch64:1406]: No :)
We need pvpanic-pci only now, I will fix that immediately.

**Comment by up2wing** [comment]
Hi all,
There are still three build check fail which I am not sure where's the problem. I revert the source code to that before I push the pvpanic patch, the problems still exist :(

**Comment by likebreath** [comment]
> Hi all, There are still three build check fail which I am not sure where's the problem. I revert the source code to that before I push the pvpanic patch, the problems still exist :(

Please rebase. I believe the issues have been fixed in the upstream main branch with #5566 and #5567.

**Comment by up2wing** [comment]
> > Hi all, There are still three build check fail which I am not sure where's the problem. I revert the source code to that before I push the pvpanic patch, the problems still exist :(
> 
> Please rebase. I believe the issues have been fixed in the upstream main branch with #5566 and #5567.

That works, thanks so much.
I will pay attention to this in the future.

File: vmm/src/device_manager.rs [hunk 7]
```diff
@@ -3604,6 +3616,44 @@ impl DeviceManager {
...
+
+        let mut node = device_node!(id, pvpanic_device);
+
+        node.resources = new_resources;
+        node.pci_bdf = Some(pci_device_bdf);
+        node.pci_device_handle = None;
+
+        self.device_tree.lock().unwrap().insert(id, node);
+
+        info!("add pvpanic device");
```
Comment (rbradford) [inline vmm/src/device_manager.rs:3653]: Please remove this log.

**Review by rbradford — APPROVED** [review]

File: vmm/src/device_manager.rs (hunk 7 above)
Comment (up2wing) [reply vmm/src/device_manager.rs]: done :)

**Pull request merged**